trie/trie.go

```diff
@@ -361,121 +345,76 @@ func (t *trie) query(key []byte) (patricia, int, error) {
-		}
-		// deleting a leaf, upstream node must be extension so collapse into extension
-		childClps, clpsType = true, 1
-	}
-	return childClps, clpsType, nil
-}
-
 // updateInsert rewinds the path back to root and updates nodes along the way
-func (t *trie) updateInsert(hashChild []byte) error {
+func (t *trie) updateInsert(curr patricia) error {
 	for t.toRoot.Len() > 0 {
-		curr, index := t.popToRoot()
-		if curr == nil {
+		next, index := t.popToRoot()
+		if next == nil || isLeaf(next) {
 			return errors.Wrap(ErrInvalidPatricia, "patricia pushed on stack is not valid")
 		}
 		// update the patricia node
-		if err := curr.ascend(hashChild[:], index); err != nil {
+		hash := curr.hash()
+		if err := next.ascend(hash[:], index); err != nil {
 			return err
 		}
-		hashCurr := curr.hash()
-		hashChild = hashCurr[:]
 		// when adding an entry, hash of nodes along the path changes and is expected NOT to exist in DB
-		if err := t.putPatricia(curr); err != nil {
+		if err := t.putPatricia(next); err != nil {
 			return err
 		}
+		curr = next
 	}
 	// update root hash
 	t.rootHash = t.root.hash()
 	return nil
 }
 
 // updateDelete rewinds the path back to root and updates nodes along the way
-func (t *trie) updateDelete(curr patricia, currClps bool, clpsType byte) error {
-	contClps := false
+func (t *trie) updateDelete() error {
+	var curr patricia
 	for t.toRoot.Len() > 0 {
 		logger.Debug().Int("stack size", t.toRoot.Len()).Msg("clps")
 		next, index := t.popToRoot()
-		if next == nil {
+		if next == nil || isLeaf(next) {
 			return errors.Wrap(ErrInvalidPatricia, "patricia pushed on stack is not valid")
 		}
 		if err := t.delPatricia(next); err != nil {
 			return errors.Wrap(err, "failed to delete patricia")
 		}
-		// we attempt to collapse in 2 cases:
-		// 1. the current node is not root
-		// 2. the current node is root, but <v> is nil meaning no more entries exist on the incoming path
-		isRoot := t.toRoot.Len() == 0
-		noEntry := t.clpsV == nil
-		var nextClps bool
-		t.clpsK, t.clpsV, nextClps = next.collapse(t.clpsK, t.clpsV, index, currClps && (!isRoot || noEntry))
-		logger.Debug().Bool("curr", currClps).Msg("clps")
-		logger.Debug().Bool("next", nextClps).Msg("clps")
-		if nextClps {
-			// current node can also collapse, concatenate the path and keep going
-			contClps = true
-			if !isRoot {
-				currClps = nextClps
-				curr = next
-				continue
-			}
-		}
-		logger.Debug().Bool("cont", contClps).Msg("clps")
-		if contClps && !noEntry {
-			curr = &leaf{clpsType, t.clpsK, t.clpsV}
-			logger.Info().Hex("k", t.clpsK).Hex("v", t.clpsV).Msg("clps")
-			// after collapsing, the trie might rollback to an earlier state in the history (before adding the deleted entry)
-			// so the node we try to put may already exist in DB
-			if err := t.putPatricia(curr); err != nil {
-				return errors.Wrap(err, "failed to put patricia")
-			}
-		}
-		contClps = false
 		// update current with new child
```

Comment:
use a simplified cleaner logic to implement updateDelete(), let me know if you have questions about details